gym/vector/async_vector_env.py

```diff
@@ -301,6 +302,88 @@ def step_wait(self, timeout=None):
             infos,
         )
 
+    def call_async(self, name, *args, **kwargs):
+        """
+        Parameters
+        ----------
+        name : string
+            Name of the method or property to call.
+        """
+        self._assert_is_running()
+        if self._state != AsyncState.DEFAULT:
+            raise AlreadyPendingCallError(
+                "Calling `call_async` while waiting "
+                "for a pending call to `{0}` to complete.".format(self._state.value),
+                self._state.value,
+            )
+
+        for pipe in self.parent_pipes:
+            pipe.send(("_call", (name, args, kwargs)))
+        self._state = AsyncState.WAITING_CALL
+
+    def call_wait(self, timeout=None):
+        """
+        Parameters
+        ----------
+        timeout : int or float, optional
+            Number of seconds before the call to `step_wait` times out. If
+            `None`, the call to `step_wait` never times out.
+
+        Returns
+        -------
+        results : list
+            List of the results of the individual calls to the method or
+            property for each environment.
+        """
+        self._assert_is_running()
+        if self._state != AsyncState.WAITING_CALL:
+            raise NoAsyncCallError(
+                "Calling `call_wait` without any prior call " "to `call_async`.",
+                AsyncState.WAITING_CALL.value,
+            )
```

Comment:
This is matching the rest of the API of `AsyncVectorEnv` (`reset_wait` and `step_wait`), but this is not a future-based one because the corresponding `*_async` methods do not return any future. The idea was to be as conservative as possible to avoid anything going wrong, but it could be swapped with a warning/empty list.

Personally, I like the option of raising an exception that can be caught by the user later on better, but I'm open to this change.

Reply:
Depends if you see AsyncVectorEnv as a state machine (in which case it may be invalid to call_wait if the environment is not in state `WAITING_CALL`, or if you see it as a parallel computing fan-out/fan-in. In the former case, I think the exception is better (we attempted an invalid action in the current state), in the latter a simple return is better (no work was scheduled, so we don't need to wait for any work to be completed).

Reply:
The idea was indeed something more like a state machine, following [the design of `SubprocVecEnv` in baselines](https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/common/vec_env/subproc_vec_env.py#L80) (which was the source of inspiration for `AsyncVectorEnv`, and our discussion in the original PR https://github.com/openai/gym/pull/1513#discussion_r291767318.

Actually looking back at the original PR, I initially had it with a simple warning, but it was suggested by the maintainers to make it fatal https://github.com/openai/gym/pull/1513#discussion_r291765948.